### PR TITLE
Add IHandSelected and IHandDeselected for the client

### PR DESCRIPTION
**Pull request by DrSmugleaf**
Missing it working when items are picked up.

### CODE REVIEW

**Comment by DrSmugleaf** [comment]
Ready for review.

**Review by metalgearsloth — APPROVED** [review]

**Comment by exp111** [comment]
Also this currently doesn't call the function when the item is picked up/dropped

File: Content.Client/UserInterface/HandsGui.cs [hunk 1]
```diff
@@ -209,7 +209,7 @@ private void HandKeyBindDown(GUIBoundKeyEventArgs args, string slotName)
 
             if (args.Function == ContentKeyFunctions.MouseMiddle)
             {
-                hands.SendChangeHand(slotName);
+                hands.SetActiveHand(slotName);
```
Comment (PJB3005) [inline Content.Client/UserInterface/HandsGui.cs:212]: You should not directly edit simulation state from UI like this. As it is this "event" is not correctly tracked by the game state system and will be undone during reconciliation. 

Use `EntityManagerExt.RaisePredictiveEvent()` to fire events into the EventBus and to the server that are correctly replayed during reconciliation. Check the combat mode GUI for an example.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Content.Client/UserInterface/HandsGui.cs (hunk 1 above)
Comment (DrSmugleaf) [reply Content.Client/UserInterface/HandsGui.cs:212]: Fixed, tell me if it's correct now.

**Comment by DrSmugleaf** [comment]
Reviews addressed.

File: Content.Client/GameObjects/Components/Items/HandsComponent.cs [hunk 2]
```diff
@@ -284,7 +356,39 @@ public Hand(HandsComponent parent, SharedHand hand, IEntityManager manager, Hand
...
+            {
+                if (_entity == value)
+                {
+                    return;
+                }
+
+                var old = _entity;
+                _entity = value;
+
+                var interactionSystem = EntitySystem.Get<SharedInteractionSystem>();
```
Comment (exp111) [inline Content.Client/GameObjects/Components/Items/HandsComponent.cs:373]: You could move that under the `if(...){return;}`

**Review by exp111 — APPROVED** [review]

File: Content.Client/GameObjects/Components/Items/HandsComponent.cs (hunk 2 above)
Comment (DrSmugleaf) [reply Content.Client/GameObjects/Components/Items/HandsComponent.cs:373]: Done.

**Comment by metalgearsloth** [comment]
@PJB3005 plz come back.

File: Content.Client/UserInterface/GameHud.cs [hunk 3]
```diff
@@ -47,6 +47,8 @@ public interface IGameHud
         Action<bool> SandboxButtonToggled { get; set; }
 
         Control HandsContainer { get; }
+        Action<string> OnHandChanged { get; set; }
```
Comment (metalgearsloth) [inline Content.Client/UserInterface/GameHud.cs:50]: Could you just invoke the eventbus directly instead? That way you can also remove the handssystem too.

File: Content.Server/GameObjects/EntitySystems/HandsSystem.cs [hunk 4]
```diff
@@ -25,7 +24,7 @@
 namespace Content.Server.GameObjects.EntitySystems
 {
     [UsedImplicitly]
-    internal sealed class HandsSystem : EntitySystem
+    internal sealed class HandsSystem : SharedHandsSystem
     {
         [Dependency] private readonly IEntityManager _entityManager = default!;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/EntitySystems/HandsSystem.cs:29]: EntityManager is already a property

File: Content.Client/GameObjects/Components/Items/HandsComponent.cs [hunk 5]
```diff
@@ -79,18 +159,16 @@ public override void Initialize()
 
         public override void HandleComponentState(ComponentState? curState, ComponentState? nextState)
         {
-            if (curState == null)
+            if (!(curState is HandsComponentState state))
```
Comment (metalgearsloth) [inline Content.Client/GameObjects/Components/Items/HandsComponent.cs:162]:
```suggestion
            if (curState is not HandsComponentState state)
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
Gonna fix conflicts and merge it in 3 days.

**Review by PaulRitter — CHANGES_REQUESTED** [review]
DNM, this uses C#-Events & has other unadressed reviews

**Comment by metalgearsloth** [comment]
image

**Pull request closed without merging**